pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -510,7 +518,14 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 //  - Write the response to the endpoint.
 func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	stat := ProxyRequestContext{}
+	stat.SemaphoreAcquireTime.Start()
+	if p.ConcurrencyLimit != nil {
+		p.ConcurrencyLimit.Acquire(context.TODO(), 1)
```

Comment:
What's the outer structure that calls `ServeDNS()`? If we receive too many DNS responses at once and start blocking here, what will the behaviour be?

Basically I'm wondering if we should really add a real context here like

```
ctx, cancel := context.WithTimeout(context.TODO(), ProxyForwardTimeout)
```

Then, we should check the returned error and potentially update `stat` & return similar to the other error handling in this function.

Reply:
^^ I put `ProxyForwardTimeout` above, which is 10 seconds by default, though I'm not sure if we should actually add a new different timeout to acquire the semaphore, like 1 second or even less. Also when returning the error, we should be sure to wrap it to add the context something like `"%s DNS requests already being handled in parallel: %w", option.DNSProxyConcurrencyLimit, err`

Reply:
@joestringer I think that adding a timeout here will counter the aim of this change - to decrease memory usage for DNS message bursts. If DNS proxy gets overflown by messages, the concurrency limit will kick in and it will limit the amount of resources that agent uses while unloading the queue.

IMO adding the timeout will just cause the memory burst to happen several seconds later, unless you want to drop messages after timeout, but this will cause DNS connectivity to drop which is also not the behaviour that we want.

Reply:
Basically current behaviour is - when DNS burst happens, we have a spike in memory usage because we are using a lot of resources to handle the DNS message. When burst ends, we unload the queue, but we are in danger of OOM while bursting.

With concurrency limiter, the memory spike doesn't really happen, because we have only `concurrencyLimit` number of dns handling data structures present and we unload the queue in much more civilized way - not allocating memory until the goroutine is able to get CPU cycles to actually handle the message.

Reply:
I was thinking about introducing the timeout and if the timeout fails, then return the error and drop the DNS request. If the rate of DNS messages coming in is too high, then we will be dropping those messages anyway, but we could make the logging/debugging/visibility more clear by explicitly handling the case, rather than implicitly causing some part of the networking stack to drop the packets instead.

My understanding of the queue is this:

* In a high load environment, the number of DNS responses per second are at some rate, say X
* The datapath and kernel will continue to feed these responses into the DNS socket at that rate, regardless of whether the cilium process keeps up or not.
* Ultimately if X is too high, then we will drop DNS messages regardless of what the Cilium process does.
* The miekg/dns library determines how to manage requests coming off the socket, probably by setting up some maximum number of goroutines/threads to handle the incoming requests, Y. (I didn't study this carefully, but I assume it's been designed in a reasonable manner like this).
* This logic in its current state will allow up to Z concurrent requests to be handled at a time.
* If the code holding the Z (this new configuration option) semaphores are delayed for some time, each goroutine/thread in Y will block at this statement for an indefinite period.
* Blocking Y handlers do not provide pressure feedback to the callers or to the user.

From here, I have two main scenarios to explore:

#### Scenario 1: Deadlock
* Worst possible case, internally in Cilium, the entire policy engine deadlocks and makes no progress. All semaphores are acquired, so no additional semaphores can be acquired in this logic.
* All Y goroutines will block since they cannot acquire a semaphore. There is no timeout, Cilium doesn't print errors to indicate the problem to the user.
* X DNS packets continue to come in, filling up the DNS socket buffer until it's full.
* The `miekg/dns` library code for handling the socket is no longer reading off the socket because all of Y are blocked.
* Somewhere around this point, the kernel socket receive logic it will start dropping packets because it can no longer enqueue the messages into the socket.
* The user is likely to observe that DNS stops working and begin investigation. I don't know what kind of errors Cilium may report here.
* If we change this so that the semaphore acquire times out after some time, then we guarantee to start printing errors in the Cilium logs that indicate that semaphores cannot be acquired and Cilium's DNS handling cannot keep up. This error message would be very useful to understand the cause and point towards where to investigate next.

#### Scenario 2: Consistently high rate X such that Cilium cannot keep up
* Consider if the rate X is higher than the rate at which Cilium can process these DNS messages with Z request handlers.
* DNS messages will come in and the Y goroutines will consume all of the X semaphores.
* The remaining Y goroutines will block while waiting for the other goroutines to process the messages.
* If the rate continues to be high, then on average, Y will not drain the incoming socket buffer quickly enough and it will eventually fill up, causing the kernel to drop the DNS messages.
* One such scenario is if Z is set too low. If Z is set too low, then we will not spend much CPU on handling DNS, and hence we'll start causing these DNS messages to be dropped simply because we don't spend enough time on DNS. In this scenario, given that Y are all blocking, we won't print a message explicitly that we're not keeping up.
  * In this scenario, we will update the DNS processing time metrics for each of the DNS messages we process, and it will be clear that we spend some amount of time on the DNS processing. But we won't print how many goroutines in Y are blocking because they can't get semaphores.
  * So I think that it will be clear that DNS message processing takes a long time, but not clear whether there is something the user can do about it.
  * If we set a timeout and print a warning, then the user can find evidence that Y is high but Z is low, and we may even be able to suggest to the user that if they just increase this new option, it may provide enough CPU to now start keeping up.
* Separately, there is also the case where Z is high and Cilium simply cannot keep up. In this case, we will  end up dropping DNS messages regardless of what we do. I suggest only that if we explicitly log that this is occurring (via this timeout), then it will help users to observe that this is the cause of the problem.

I note that if these goroutines do not block, then they will discard the DNS message and then go and fetch another message off the socket to try to process it. This will take _some_ more CPU, hard to say how much. At the same time, this will also continue to provide feedback to the user that this rate X is too high, so the issue should continue to be fairly visible during the incident.

I should also note that I've been thinking about the above in terms of logging an error. We could alternatively (or additionally!) try to provide this feedback in terms of metrics.

#### Side note on queueing and providing pressure feedback to callers

I also had some other notions about a scenario where the X rate oscillates between X being "too many to keep up" and "can keep up with this many", but after a few minutes of consideration I don't think that it will make a meaningful enough difference right now to discuss further. Broadly I was thinking about how buffers cause a delay in providing signal back to callers, so if we can reduce that delay and signal to callers (ie original apps) that the rate is too high, then we could potentially assist in lowering the rate X. But I don't think we have enough structure in place today to really solve that sort of problem. A timeout would be required to solve that problem, but we'd also need to do further development work which is not currently on the roadmap.

Reply:
Resolved in https://github.com/cilium/cilium/pull/19592#issuecomment-1121971577